pkg/controller/controller.go

```diff
@@ -168,90 +168,103 @@ func (c *Controller) GetLastErrorTimestamp() time.Time {
 func (c *Controller) runController() {
 	errorRetries := 1
 
-	// ensure the callbacks are valid
-	if c.params.DoFunc == nil {
-		c.params.DoFunc = func() error { return undefinedDoFunc(c.name) }
-	}
-	if c.params.StopFunc == nil {
-		c.params.StopFunc = NoopFunc
-	}
+	c.mutex.RLock()
+	params := c.params
+	c.mutex.RUnlock()
+	runFunc := true
+	interval := 10 * time.Minute
 
 	for {
-		var (
-			err      error
-			interval = c.params.RunInterval
-		)
-
-		err = c.params.DoFunc()
-		if err != nil {
-			switch err := err.(type) {
-			case ExitReason:
-				// This is actually not an error case, but it causes an exit
-				c.recordSuccess()
-				c.lastError = err // This will be shown in the controller status
-				goto shutdown
-
-			default:
-				c.getLogger().WithField(fieldConsecutiveErrors, errorRetries).
-					WithError(err).Debug("Controller run failed")
-				c.recordError(err)
-
-				if !c.params.NoErrorRetry {
-					if c.params.ErrorRetryBaseDuration != time.Duration(0) {
-						interval = time.Duration(errorRetries) * c.params.ErrorRetryBaseDuration
-					} else {
-						interval = time.Duration(errorRetries) * time.Second
+		var err error
+		if runFunc {
+			interval = params.RunInterval
+
+			err = params.DoFunc()
+
+			c.mutex.Lock()
+
+			if err != nil {
+				switch err := err.(type) {
+				case ExitReason:
```

Comment:
We ended up not using this anywhere. If it makes your PRs simpler, we can remove the exit behaviour outright.